core/system/src/page_allocator.c

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2013-2015, ARM Limited, All Rights Reserved
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <uvisor.h>
+#include "page_allocator.h"
+
+
+/* Must be UVISOR_PAGE_SIZE for MPU alignment in ARMv7-M */
+#define UVISOR_PAGE_ALIGNMENT UVISOR_PAGE_SIZE
+
+/* We can only have a total of 64 pages for now */
+#define UVISOR_PAGE_TABLE_COUNT ((size_t)64)
```

Comment:
Hm, currently it's not completely implementation independent, in `page_free` there is this (ugly) bit mask:
```
 /* contains the bit mask of all the pages we need to free */
uint64_t free_mask = 0;
```
The idea was that you only compute the pages to be freed once, and store it in a bitmask.
I can rewrite it to compute the page index on the fly again from the page origin, that would make this implementation independent.